nengo/connection.py

```diff
@@ -36,24 +36,35 @@ def __set__(self, conn, rule):
 
     def validate_rule(self, conn, rule):
         super(ConnectionLearningRuleTypeParam, self).validate_rule(conn, rule)
-        rule_type = 'decoders'
-        if conn.solver.weights or (isinstance(conn.pre_obj, Neurons)
-                                   and isinstance(conn.post_obj, Neurons)):
-            rule_type = 'transform'
 
-        if rule_type not in rule.modifies:
-            raise ValueError("Learning rule '%s' cannot be applied to "
-                             "a %s-based connection." % (rule, rule_type))
-
-        if not isinstance(conn.pre_obj, (Ensemble, Neurons)):
-            raise ValueError("'pre' must be of type 'Ensemble' or 'Neurons' "
-                             "for learning rule '%s' (got type '%s')" % (
-                                 rule, conn.pre_obj.__class__.__name__))
+        # --- Check pre object
+        if rule.modifies in ('decoders', 'weights'):
+            # pre object must be neural
+            if not isinstance(conn.pre_obj, (Ensemble, Neurons)):
+                raise ValueError(
+                    "'pre' must be of type 'Ensemble' or 'Neurons' for "
+                    "learning rule '%s' (got type '%s')" % (
+                        rule, conn.pre_obj.__class__.__name__))
+
+        # --- Check post object
+        if rule.modifies == 'encoders':
+            if not isinstance(conn.post_obj, Ensemble):
+                raise ValueError("'post' must be of type 'Ensemble' (got %r) "
+                                 "for learning rule '%s'"
+                                 % (conn.pre_obj.__class__.__name__, rule))
+        else:
+            if not isinstance(conn.post_obj, (Ensemble, Neurons, Node)):
+                raise ValueError(
+                    "'post' must be of type 'Ensemble', 'Neurons' or 'Node' "
+                    "(got %r) for learning rule '%s'"
+                    % (conn.post_obj.__class__.__name__, rule))
 
-        if not isinstance(conn.post_obj, (Ensemble, Neurons, Node)):
-            raise ValueError("'post' must be of type 'Ensemble', 'Neurons' "
-                             "or 'Node' for learning rule '%s' (got type '%s')"
-                             % (rule, conn.post_obj.__class__.__name__))
+        # If the rule modifies 'weights', then it must have full weights
+        if rule.modifies == 'weights' and conn.is_factored:
+            raise ValueError(
+                "Learning rule '%s' can not be applied to factored "
```

Comment:
Again I think 'factored' is a confusing term. I think in the past we've used "decoded" connection versus "direct" or "neuron to neuron" connections, but I agree that those aren't super clear either...